firedrake/linear_solver.py

```diff
@@ -41,153 +31,59 @@ def __init__(self, A, *, P=None, solver_parameters=None,
                created.  Use this option if you want to pass options
                to the solver from the command line in addition to
                through the ``solver_parameters`` dict.
-        :kwarg pre_apply_bcs: Ignored by this class.
+        :kwarg pre_apply_bcs: If `True`, the bcs are applied before the solve.
+               Otherwise, the bcs are included as part of the linear system.
 
         .. note::
 
           Any boundary conditions for this solve *must* have been
           applied when assembling the operator.
         """
-        if not isinstance(A, matrix.MatrixBase):
+        if not isinstance(A, MatrixBase):
             raise TypeError("Provided operator is a '%s', not a MatrixBase" % type(A).__name__)
-        if P is not None and not isinstance(P, matrix.MatrixBase):
+        if P is not None and not isinstance(P, MatrixBase):
             raise TypeError("Provided preconditioner is a '%s', not a MatrixBase" % type(P).__name__)
 
-        solver_parameters = flatten_parameters(solver_parameters or {})
-        solver_parameters = solving_utils.set_defaults(solver_parameters,
-                                                       A.arguments(),
-                                                       ksp_defaults=self.DEFAULT_KSP_PARAMETERS)
+        test, trial = A.arguments()
+        self.x = Function(trial.function_space())
+        self.b = Cofunction(test.function_space().dual())
+
+        problem = LinearVariationalProblem(A, self.b, self.x, aP=P,
+                                           form_compiler_parameters=A.fc_params,
```

Comment:
Let's help to anyone reading the code and be a bit more explicit - this is what is used essentially everywhere except the `ImplicitMatrixContext`.
```suggestion
                                           form_compiler_parameters=A.form_compiler_parameters,
```